modules/katamari/src/main/ts/ephox/katamari/api/Maybes.ts

```diff
@@ -0,0 +1,327 @@
+} = <T>(predicate: (value: T) => boolean) => (self: Maybe<T>) => {
+  if (isJust(self) && predicate(self.value)) {
+    return just(self.value);
+  } else {
+    return nothing();
+  }
+};
+
+// --- Getters ---
+
+// Get the value out of the inside of the Maybe object, using a default
+// `replacement` value if the provided Maybe object does not contain a value.
+export const getOr = <T, U = T>(replacement: U) => (self: Maybe<T>): T | U =>
+  isJust(self) ? self.value : replacement;
+
+// Get the value out of the inside of the Maybe object, throwing an exception if
+// the provided Maybe object does not contain a value.
+//
+// WARNING:
+// 1. You should only be using this function if you know that the Maybe object
+// **is not** empty (otherwise you're throwing exceptions in production code,
+// which is bad).
+// 2. If you know that the Maybe object is not empty, then TypeScript should
+// also know that the Maybe object is not empty (comparison operations on `.tag`
+// as well as the `isJust` method will both allow the compiler to infer that the
+// object is not empty).
+// 3. If TypeScript knows that the Maybe object is not empty, it will let you
+// just access `.value` directly, without needing this function + without
+// risking an exception being thrown.
+//
+// Therefore, you should only be calling this function in environments where
+// you're unable to completely type-check your code (IE: tests). In all other
+// situations, **avoid this method**.
+export const getOrDie = <T>(self: Maybe<T>): T => {
+  if (isNothing(self)) {
+    throw new Error('Called getOrDie on Nothing');
+  }
+
+  return self.value;
+};
+
+// Get the value out of the inside of the Maybe object, using a default
+// `replacement` value if the provided Maybe object does not contain a value.
+// Unlike `getOr`, in this method the `replacement` value is "thunked" - that is
+// to say that you don't pass a value to `getOrThunk`, you pass a function which
+// (if called) will **return** the `value` you want to use.
+export const getOrThunk = <T, U = T>(thunk: () => U) => (self: Maybe<T>): T | U =>
+  isJust(self) ? self.value : thunk();
+
+// --- Comparators ---
+
+// **Is** the value stored inside this Maybe object equal to `other`? Uses `===`
+// to check.
+export const is = <T>(other: T) => (self: Maybe<T>): boolean =>
+  isJust(self) && self.value === other;
+
+// Are these two Maybe objects equal? Equality here means either they're both
+// `Just` (and the values are equal under `===`) or they're both `Nothing`.
+export const equals = <T>(lhs: Maybe<T>, rhs: Maybe<T>): boolean => {
+  if (isJust(lhs) && isJust(rhs)) {
+    return lhs.value === rhs.value;
+  } else {
+    return lhs.tag === rhs.tag;
+  }
+};
+
+// Same as `equals`, but instead of using `===` a custom comparator function is used.
+export const equals_ = <T, U>(lhs: Maybe<T>, rhs: Maybe<U>, comparator: (lhs: T, rhs: U) => boolean): boolean => {
```

Comment:
I wonder if we should take the chance to rename this to something saner? Maybe `compare` or something similar instead?

Reply:
Can we just add the comparator as a defaulted param to `equals`?

Reply:
Actually, yes that would be a better idea :+1: We can just use `Fun.tripleEquals` as the default comparator.

Reply:
Done, please double check I didn't ruin the typescript while doing this though. The type signatures got a little weird

Reply:
LGTM